nixos/modules/services/torrent/deluge.nix

```diff
@@ -15,50 +39,160 @@ in {
+            boolean values must not. See
+            https://git.deluge-torrent.org/deluge/tree/deluge/core/preferencesmanager.py#n41
+            for the availaible options.
+          '';
+        };
+
+        declarative = mkOption {
+          type = types.bool;
+          default = false;
+          description = ''
+            Whether to use a declarative deluge configuration.
+            Only if set to <literal>true</literal>, the options
+            <option>services.deluge.config</option>,
+            <option>services.deluge.openFirewall</option> and
+            <option>services.deluge.authFile</option> will be
+            applied.
+          '';
+        };
+
+        openFirewall = mkOption {
+          default = false;
+          type = types.bool;
+          description = ''
+            Whether to open the firewall for the ports in
+            <option>services.deluge.config.listen_ports</option>. It only takes effet if
+            <option>services.deluge.declarative</option> is set to
+            <literal>true</literal>.
+
+            It does NOT apply to the daemon port nor the web UI port. To access those
+            ports secuerly check the documentation
+            https://dev.deluge-torrent.org/wiki/UserGuide/ThinClient#CreateSSHTunnel
+            or use a VPN or configure certificates for deluge.
+          '';
+        };
+
+        dataDir = mkOption {
+          type = types.path;
+          default = "/var/lib/deluge";
+          description = ''
+            The directory where deluge will create files.
+          '';
+        };
+
+        authFile = mkOption {
+          type = types.path;
+          example = "/run/keys/deluge-auth";
+          description = ''
+            The file managing the authentication for deluge, the format of this
+            file is straightforward, each line contains a
+            username:password:level tuple in plaintext. It only has an effect
+            when <option>services.deluge.declarative</option> is set to
+            <literal>true</literal>.
+            See https://dev.deluge-torrent.org/wiki/UserGuide/Authentication for
+            more informations.
+          '';
+        };
       };
 
-      deluge.web.enable = mkEnableOption "Deluge Web daemon";
+      deluge.web = {
+        enable = mkEnableOption "Deluge Web daemon";
+        port = mkOption {
+        type = types.port;
+          default = 8112;
+          description = ''
+            Deluge web UI port.
+          '';
+        };
+      };
     };
   };
 
   config = mkIf cfg.enable {
 
+    systemd.tmpfiles.rules = [ "d '${configDir}' 0770 deluge deluge" ]
+    ++ optional (hasAttr "download_location" cfg.config)
+      "d '${cfg.config.download_location}' 0770 deluge deluge"
+    ++ optional (hasAttr "torrentfiles_location" cfg.config)
+      "d '${cfg.config.torrentfiles_location}' 0770 deluge deluge"
+    ++ optional (hasAttr "move_completed_path" cfg.config)
```

Comment:
And something minor here too: `hasAttr "name" value` is equivalent to `value ? name`, so you can use `optional (cfg.config ? move_completed_path)` and such here.